File: app/src/main/cpp/02CPP/_20140822/ExternTest.java
/**
 * Created by along on 17/3/13 14:48.
 * Email:466210864@qq.com
 * java中：
 * protected 修饰的成员变量和函数能被类本身、子类及同一个包中的类访问。
 */
public class ExternTest extends Father {
    public static void main(String[] args) {
        Father father = new Father();
        father.callInClass();
        father.proMethod();
        System.out.println();

        ProSon proSon = new ProSon();
        proSon.callInClass();
        proSon.proMethod();
        System.out.println();

        Father fa = new ProSon();
        fa.proMethod();//这里与 C++不同，C++这里调用的是父类的方法，估计这里底层使用了 C++的virtual，但这里如果修改使其能调用父类的方法？
        fa.callInClass();
        Father fa2 = fa;
    }

}

class Father {
    public Father() {
        priChar = "F priChar";
        protChar = "F proChar";
        pubChar = "F pubChar";
        System.out.println("Father 构造方法");
    }

    private String priChar;
    protected String protChar;

    protected void proMethod() {
        System.out.println("父类的 protected 方法");
    }

    public String pubChar;


    public void callInClass() {
        System.out.println("父类的公有方法");
    }
}

class ProSon extends Father {
    public ProSon() {
//        priChar = "F priChar";
        protChar = "S proChar";
        pubChar = "S pubChar";
        System.out.println("ProSon 构造方法");
    }
    protected void proMethod() {
        System.out.println("子类的 protected 方法");
    }

    public String pubChar;


    public void callInClass() {
        System.out.println("子类的公有方法");
    }
}

File: app/src/main/cpp/02CPP/_20140822/externTest.cpp
//
// Created by sunalong on 17/3/13 14:12.
// Email:466210864@qq.com
//
#include<iostream>
#include<stdlib.h>

using namespace std;

class Father {
private:
    char *priChar;
protected:
    char *protChar;
    void proMethod() {
        cout << "父类的 protected 方法" << endl;
    }
public:
    char *pubChar;

    Father() {
        priChar = new char;
        priChar = "F priChar";
        protChar = new char;
        protChar = "F proChar";
        pubChar = new char;
        pubChar = "F pubChar";
        cout << "Father()" << endl;
    }

    ~Father() {
        cout << "~Father()" << endl;
    }

    void callInClass() {
        cout << "父类的公有方法"<<endl;
    }
};

class PubSon : public Father {
public:
    PubSon() {
        cout << "PubSon()" << endl;
    }

    ~PubSon() {
        cout << "~PubSon()" << endl;
    }

    void callInClass() {
        cout << "子类覆盖父类的公有方法"<<endl;
    }
};

class ProSon : protected Father {
public:
    ProSon() {
        cout << "ProSon()" << endl;
    }

    ~ProSon() {
        cout << "~ProSon()" << endl;
    }

    void callInClass() {
        cout << "子类覆盖父类的公有方法"<<endl;
    }
};

int main() {
    Father father;
    father.callInClass();
//    father.proMethod();
    cout << "protect 方法在类外部无法调用，只能本类调用" << endl << endl;

    PubSon pubSon;
    pubSon.callInClass();
    pubSon.Father::callInClass();
    cout << "public 继承属于原样继承，不改变父类在子类中的属性" << endl << endl;

    ProSon proSon;
    proSon.callInClass();
//    proSon.Father::callInClass();
    cout<<"protected 继承会将原 public 成员改成 protected,这样父类的 public 方法无法在外部子类外部被调用"<<endl<<endl;
    system("rm -rf a.out*");
    return 0;
}

